File: AppStoreReviewTime/RTTwitterService.m
//
//  RTTwitterService.m
//  AppStoreReviewTime
//
//  Created by Lukas Würzburger on 6/7/16.
//  Copyright © 2016 Truffls GmbH. All rights reserved.
//

#import "RTTwitterService.h"

#import <Accounts/Accounts.h>
#import <Social/Social.h>

@interface RTTwitterService () {
    NSMutableDictionary *stateDictionary;
}

@property (nonatomic, strong) ACAccountStore *accountStore;

@end

@implementation RTTwitterService

+ (instancetype)sharedService {
    
    static RTTwitterService *service;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        service = [[RTTwitterService alloc] init];
    });
    return service;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        
        stateDictionary = [self currentStateFile];
        
        self.accountStore = [[ACAccountStore alloc] init];
        ACAccountType *accountType = [self.accountStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
        
        [self.accountStore requestAccessToAccountsWithType:accountType
                                         options:nil
                                      completion:^(BOOL granted, NSError *error) {
            if (granted == YES) {
                NSLog(@"Yes");
            } else {
                NSLog(@"No %@", error);
            }
        }];
    }
    return self;
}

- (NSMutableDictionary *)currentStateFile {
    
    NSString *homeDirectory = NSHomeDirectory();
    NSString *fileName = @".asrt-watcher";
    NSString *filePath = [NSString stringWithFormat:@"%@/%@", homeDirectory, fileName];
    if ([[NSFileManager defaultManager] fileExistsAtPath:filePath]) {
        return [NSMutableDictionary dictionaryWithContentsOfFile:filePath];
    } else {
        return [NSMutableDictionary dictionary];
    }
}

- (NSArray *)accounts {

    ACAccountType *accountType = [self.accountStore accountTypeWithAccountTypeIdentifier:ACAccountTypeIdentifierTwitter];
    return [self.accountStore accountsWithAccountType:accountType];
}

- (NSDate *)dateOfLastTweet {
    
    return [stateDictionary objectForKey:@"lastTweetDate"];
}

- (void)saveState {
    
    NSString *homeDirectory = NSHomeDirectory();
    NSString *fileName = @".asrt-watcher";
    NSString *filePath = [NSString stringWithFormat:@"%@/%@", homeDirectory, fileName];
    BOOL s = [stateDictionary writeToFile:filePath atomically:NO];
    NSLog(@"%i", s);
}

- (BOOL)isAccountConnected {
    
    NSArray *accounts = [self accounts];
    NSLog(@"%@", accounts);
    if (accounts.count > 0) {
        
        return YES;
    }
    return NO;
}

- (void)post:(NSString *)tweet block:(void (^)(BOOL success))block {
    
    NSDictionary *message = @{ @"status" : tweet };
    
    NSURL *requestURL = [NSURL URLWithString:@"https://api.twitter.com/1.1/statuses/update.json"];
    
    ACAccount *twitterAccount = [self accounts].lastObject;
    
    SLRequest *postRequest = [SLRequest requestForServiceType:SLServiceTypeTwitter requestMethod:SLRequestMethodPOST URL:requestURL parameters:message];
    postRequest.account = twitterAccount;
    
    [postRequest performRequestWithHandler:^(NSData *responseData, NSHTTPURLResponse *urlResponse, NSError *error) {
        NSLog(@"Twitter HTTP response: %li", (long)[urlResponse statusCode]);
        
        BOOL success = (error == nil);
        if (success) {
        
            [stateDictionary setObject:[NSDate date] forKey:@"lastTweetDate"];
            [self saveState];
        }
        
        if (block) {
            block(success);
        }
    }];
}

@end

File: AppStoreReviewTime/RTTwitterService.swift
//
//  RTTwitterService.swift
//  AppStoreReviewTime
//
//  Created by Lukas on 11/16/16.
//  Copyright © 2016 Truffls GmbH. All rights reserved.
//

import Cocoa
import Accounts
import Social

class RTTwitterService {

    // MARK: - Variables
    
    var accountStore: ACAccountStore!
    
    // MARK: - Initialization
    
    init() {
        
        accountStore = ACAccountStore()
        accountStore.requestAccessToAccounts(with: twitterAccountType(), options: [:]) { (granted, error) in
            print("Granted: \(granted) Error: \(error)")
        }
    }
    
    func twitterAccountType() -> ACAccountType {
        return accountStore.accountType(withAccountTypeIdentifier: ACAccountTypeIdentifierTwitter)
    }
    
    func dateOfLastTweet() -> Date? {
        var date: Date?
        let filePath = "\(NSHomeDirectory())/.asrt-watcher"
        if FileManager.default.fileExists(atPath: filePath) {
            if let file = NSDictionary(contentsOfFile: filePath) as? [String : Date] {
                date = file["lastTweetDate"]
            }
        }
        return date
    }
    
    func saveDate(_ date: Date) {
        let filePath = "\(NSHomeDirectory())/.asrt-watcher"
        let dict = NSDictionary(dictionary: ["lastTweetDate" : date])
        dict.write(toFile: filePath, atomically: false)
    }
    
    func post(_ message: String, block: @escaping (_ success: Bool) -> Void) {
        let tweet = ["status" : message]
        let url = URL(string: "https://api.twitter.com/1.1/statuses/update.json")
        let account = self.accountStore.accounts(with: twitterAccountType()).last
        let request = SLRequest(forServiceType: SLServiceTypeTwitter, requestMethod: .POST, url: url, parameters: tweet)
        request?.account = account as! ACAccount!
        request?.perform(handler: { (responseData, urlResponse, error) in
            
            if error == nil {
                self.saveDate(Date())
            }
            block(error == nil)
        })
    }
}
